### PR TITLE
MCKIN-14035 Optimized course tree API

### PR DESCRIPTION
**Pull request by nasirhjafri**
- Removed unnecessary nested loop.
- Removed `.copy()` on blocks.
- Moved logic to find course block to a new function to avoid iterating all of the blocks

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @nasirhjafri! I've created [OSPR-4273](https://openedx.atlassian.net/browse/OSPR-4273) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://open.edx.org/wp-content/uploads/2019/01/individual-contributor-agreement.pdf)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by nasirhjafri** [comment]
Thanks, @naeem91 for reviewing :)

File: edx_solutions_api_integration/courses/views.py [hunk 1]
```diff
@@ -2659,24 +2659,25 @@ def post(self, request):
...
-                    "id": str(course_id),
-                    "content": course['children']
-                }
-                response_data.append(course_data)
+            blocks = course_structure.structure.get('blocks', {})
+            for block in blocks.values():
+                block['name'] = block.pop('display_name')
+                block['category'] = block.pop('block_type')
+
+            course = self._get_course(blocks)
```
Comment (ihtram) [inline edx_solutions_api_integration/courses/views.py]:
```
for block in blocks.values():
    block['name'] = block.pop('display_name')
    block['category'] = block.pop('block_type')
course = self._get_course(blocks)
```

This code block can be optimized to:

```
course = None
for block in blocks.values():
    block['name'] = block.pop('display_name')
   block['category'] = block.pop('block_type')
    if block['category'] == 'course' and course == None:
        course = block
```

@nasirhjafri what do you say?

File: edx_solutions_api_integration/courses/views.py [hunk 2]
```diff
@@ -2659,24 +2659,25 @@ def post(self, request):
...
-                    block['category'] = block.pop('block_type')
-
-                course = [block for block_id, block in blocks.items() if block['category'] == 'course'][0]
-                self._update_blocks(course, blocks)
-                course_data = {
-                    "id": str(course_id),
-                    "content": course['children']
-                }
-                response_data.append(course_data)
+            blocks = course_structure.structure.get('blocks', {})
```
[the remark this answers is not in the document]
Comment (nasirhjafri) [reply edx_solutions_api_integration/courses/views.py]: @ihtram It was taking a few ms, also It wasn't needed.

File: edx_solutions_api_integration/courses/views.py (hunk 1 above)
Comment (nasirhjafri) [reply edx_solutions_api_integration/courses/views.py]: @ihtram sounds good, lemme do a quick PR.

File: edx_solutions_api_integration/courses/views.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ihtram) [reply edx_solutions_api_integration/courses/views.py]: @nasirhjafri removing .copy() will result in making changes in the original object. If there is no impact of changing the original object and it is saving some time then we can remove it.

**Pull request merged**